src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -0,0 +1,1365 @@
+                    ConstructorUsed = "int";
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list = [with(42), 1];
+                    Console.WriteLine(list.ConstructorUsed);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("int"));
+    }
+
+    [Fact]
+    public void WithElement_NoMatchingConstructor()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public MyList(string value) : base() { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    MyList<int> list = [with(42)];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics(
+            // (12,34): error CS1503: Argument 1: cannot convert from 'int' to 'string'
+            //         MyList<int> list = [with(42)];
+            Diagnostic(ErrorCode.ERR_BadArgType, "42").WithArguments("1", "int", "string").WithLocation(12, 34));
+    }
+
+    #endregion
+
+    #region Accessibility Tests
+
+    [Fact]
+    public void WithElement_PrivateConstructor()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                private MyList(int capacity) : base(capacity) { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    MyList<int> list = [with(10)];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics(
+            // (12,28): error CS1729: 'MyList<int>' does not contain a constructor that takes 0 arguments
+            //         MyList<int> list = [with(10)];
+            Diagnostic(ErrorCode.ERR_BadCtorArgCount, "[with(10)]").WithArguments("MyList<int>", "0").WithLocation(12, 28));
+    }
+
+    [Fact]
+    public void WithElement_ProtectedConstructor()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                protected MyList(int capacity) : base(capacity) { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    MyList<int> list = [with(10)];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics(
+            // (12,28): error CS1729: 'MyList<int>' does not contain a constructor that takes 0 arguments
+            //         MyList<int> list = [with(10)];
+            Diagnostic(ErrorCode.ERR_BadCtorArgCount, "[with(10)]").WithArguments("MyList<int>", "0").WithLocation(12, 28));
```

Comment:
Consider testing a version of this in a sublcass of `MyList`

Reply:
bab3841a2a9